### PR TITLE
[tune] Add config logging functionality to PBT scheduler

### PR DESCRIPTION
**Pull request by arcelien**
## What do these changes do?
Logs additional files to ray local (results) dir, to keep track of PBT explore/exploit history. This allows the PBT schedule to be reconstructed after PBT has finished.

Should double check that method for getting trial name/id is compatible with "reuse_actors=True".

## Linter

- [ x ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/tune/schedulers/pbt.py [hunk 1]
```diff
@@ -161,7 +168,8 @@ def __init__(self,
                  perturbation_interval=60.0,
                  hyperparam_mutations={},
                  resample_probability=0.25,
-                 custom_explore_fn=None):
+                 custom_explore_fn=None,
+                 log_config=False):
```
Comment (ericl) [inline python/ray/tune/schedulers/pbt.py]: We can always do this by default? That would avoid the need for a flag.

File: python/ray/tune/schedulers/pbt.py [hunk 2]
```diff
@@ -228,6 +239,34 @@ def _exploit(self, trial_executor, trial, trial_to_clone):
                     "{} (score {}) -> {} (score {})".format(
                         trial_to_clone, new_state.last_score, trial,
                         trial_state.last_score))
+
+        if self._log_config:
```
Comment (ericl) [inline python/ray/tune/schedulers/pbt.py]: Can we make this a helper method?

**Review by ericl — COMMENTED** [review]
Can you also add a unit test in trial_scheduler_test?

File: python/ray/tune/schedulers/pbt.py (hunk 1 above)
Comment (arcelien) [reply python/ray/tune/schedulers/pbt.py]: Do you mean we should always do it and remove the flag, or have it set to a default_true setting?

File: python/ray/tune/tests/test_trial_scheduler.py [hunk 3]
```diff
@@ -578,7 +568,7 @@ class _MockTrial(Trial):
     def __init__(self, i, config):
         self.trainable_name = "trial_{}".format(i)
         self.config = config
-        self.experiment_tag = "tag"
+        self.experiment_tag = "{}tag".format(i)
```
Comment (arcelien) [inline python/ray/tune/tests/test_trial_scheduler.py]: Are there any limitations on the tag being the trial id initially? My code assumes that the tag is an integer (trial id), which can have @perturbed appended to it. I think this holds in the general use case, but not for these tests. I can change it to be more generic.

File: python/ray/tune/tests/test_trial_scheduler.py [hunk 4]
```diff
@@ -889,6 +865,47 @@ def testPerturbationResetsLastPerturbTime(self):
         pbt.on_trial_result(runner, trials[3], result(11000, 100))
         self.assertEqual(pbt._num_perturbations, 2)
 
+    def testLogConfig(self):
```
Comment (arcelien) [inline python/ray/tune/tests/test_trial_scheduler.py]: I tried to write this in the style of the surrounding tests, but would be great if you can double check especially the tmpdir and file reading code.

**Review by ericl — APPROVED** [review]
LGTM. I pushed two changes:
- prefix files with pbt_
- write lines using json.dumps() instead of py ast

**Comment by arcelien** [comment]
Thanks Eric!

**Comment by ericl** [comment]
Ah oops, it looks like the py2 travis tests are actually failing here. This is because the strings are loaded as unicode type.

**Comment by ericl** [comment]
Fixed in https://github.com/ray-project/ray/pull/4710

**Pull request merged**